### PR TITLE
[Keyboard] Add Corin Keyboard

### PR DESCRIPTION
**Pull request by seatonjiang**
## Description

Corin is a 64-key keyboard that adds 4 more arrow keys than a regular 60% keyboard, and has Home and End buttons next to the arrow keys.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/longnald/corin/config.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    #define RGBLIGHT_SAT_STEP 8
+    #define RGBLIGHT_VAL_STEP 8
+    #define RGBLIGHT_LIMIT_VAL 255
+    #define RGBLIGHT_SLEEP
+    #define RGBLIGHT_EFFECT_BREATHING
+    #define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+    #define RGBLIGHT_EFFECT_SNAKE
+    #define RGBLIGHT_EFFECT_CHRISTMAS
+    #define RGBLIGHT_EFFECT_STATIC_GRADIENT
+#endif
```
Comment (drashna) [inline keyboards/longnald/corin/config.h:53]:
```suggestion
#ifdef RGB_DI_PIN
#    define RGBLED_NUM 64
#    define RGBLIGHT_HUE_STEP 8
#    define RGBLIGHT_SAT_STEP 8
#    define RGBLIGHT_VAL_STEP 8
#    define RGBLIGHT_LIMIT_VAL 255
#    define RGBLIGHT_SLEEP
#    define RGBLIGHT_EFFECT_BREATHING
#    define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#    define RGBLIGHT_EFFECT_SNAKE
#    define RGBLIGHT_EFFECT_CHRISTMAS
#    define RGBLIGHT_EFFECT_STATIC_GRADIENT
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/longnald/corin/config.h (hunk 1 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/config.h]: This has been fixed in a new commit, thanks for the correction!

File: keyboards/longnald/corin/rules.mk [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+F_CPU = 16000000
+
+# LUFA specific
+ARCH = AVR8
+
+# Input clock frequency
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
```
[the remark this answers is not in the document]
Comment (seatonjiang) [reply keyboards/longnald/corin/rules.mk]: This has been fixed in a new commit, thanks for the correction!

File: keyboards/longnald/corin/config.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x5633
+#define PRODUCT_ID      0x1024
+#define DEVICE_VER      0x0004
+#define MANUFACTURER    Longnald
+#define PRODUCT         Longnald Corin Keyboard
```
Comment (fauxpark) [inline keyboards/longnald/corin/config.h:28]:
```suggestion
#define PRODUCT         Corin
```

File: keyboards/longnald/corin/info.json [hunk 4]
```diff
@@ -0,0 +1,78 @@
+{
+  "keyboard_name": "Corin",
+  "url": "https://github.com/longnald/corin",
+  "maintainer": "Seaton Jiang <seaton@vtrois.com>",
+  "width": 28.54,
+  "height": 9.58,
+  "layouts": {
+    "LAYOUT": {
+      "key_count": 64,
```
Comment (fauxpark) [inline keyboards/longnald/corin/info.json:9]:
```suggestion
```

File: keyboards/longnald/corin/readme.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+![Corin view](https://github.com/longnald/corin/raw/master/view/view-1.jpg)
+
+Corin is a 64-key keyboard that adds 4 more arrow keys than a regular 60% keyboard made and sold by Seaton Jiang.
+
+Keyboard Maintainer: [Longnald](https://github.com/longnald) & [Seaton Jiang](https://github.com/seatonjiang)
+
+Hardware Supported: [Corin PCB (uses 32U4)](https://github.com/longnald/corin)
+
+Hardware Availability: [wwww.longnald.com](https://wwww.longnald.com/)
```
Comment (fauxpark) [inline keyboards/longnald/corin/readme.md:11]:
```suggestion
* Keyboard Maintainer: [Longnald](https://github.com/longnald) & [Seaton Jiang](https://github.com/seatonjiang)
* Hardware Supported: [Corin PCB (uses 32U4)](https://github.com/longnald/corin)
* Hardware Availability: [wwww.longnald.com](https://wwww.longnald.com/)
```

File: keyboards/longnald/corin/readme.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+Make example for this keyboard (after setting up your build environment):
+
+```shell
+make longnald/corin:default
+```
```
Comment (fauxpark) [inline keyboards/longnald/corin/readme.md:17]:
```suggestion
    make longnald/corin:default
```

File: keyboards/longnald/corin/rules.mk [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
```
Comment (fauxpark) [inline keyboards/longnald/corin/rules.mk:16]:
```suggestion
```

File: keyboards/longnald/corin/rules.mk [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
```
Comment (fauxpark) [inline keyboards/longnald/corin/rules.mk:44]:
```suggestion
# Build Options
#   change yes to no to disable
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/longnald/corin/config.h (hunk 3 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/config.h:28]: done

File: keyboards/longnald/corin/info.json (hunk 4 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/info.json:9]: done

File: keyboards/longnald/corin/readme.md (hunk 5 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/readme.md:11]: done

File: keyboards/longnald/corin/readme.md (hunk 6 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/readme.md:17]: done

File: keyboards/longnald/corin/rules.mk (hunk 7 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/rules.mk:16]: done

File: keyboards/longnald/corin/rules.mk (hunk 8 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/rules.mk:44]: done

**Comment by seatonjiang** [comment]
@drashna @fauxpark  Thanks for the correction, the above question has been corrected.

**Review by fauxpark — APPROVED** [review]

File: keyboards/longnald/corin/rules.mk [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+BOOTLOADER = atmel-dfu
+
+# Build Options
+#   change yes to no to disable
```
Comment (fauxpark) [inline keyboards/longnald/corin/rules.mk:8]:
```suggestion
#   change yes to no to disable
#
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/longnald/corin/info.json [hunk 10]
```diff
@@ -0,0 +1,77 @@
+{
+  "keyboard_name": "Corin",
+  "url": "https://github.com/longnald/corin",
+  "maintainer": "Seaton Jiang <seaton@vtrois.com>",
+  "width": 28.54,
+  "height": 9.58,
```
Comment (noroadsleft) [inline keyboards/longnald/corin/info.json:6]:
```suggestion
  "width": 15,
  "height": 5,
```

File: keyboards/longnald/corin/corin.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (noroadsleft) [inline keyboards/longnald/corin/corin.h:22]:
```suggestion
#define LAYOUT_64_ansi( \
```

File: keyboards/longnald/corin/info.json [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  "layouts": {
+    "LAYOUT": {
```
Comment (noroadsleft) [inline keyboards/longnald/corin/info.json:8]:
```suggestion
    "LAYOUT_64_ansi": {
```

File: keyboards/longnald/corin/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    [0] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/longnald/corin/keymaps/default/keymap.c:21]:
```suggestion
    [0] = LAYOUT_64_ansi(
```

File: keyboards/longnald/corin/keymaps/default/keymap.c [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        KC_ESC, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, KC_9, KC_0, KC_MINS, KC_EQL, KC_BSPC,
+        KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, KC_O, KC_P, KC_LBRC, KC_RBRC, KC_BSLS,
+        KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, KC_L, KC_SCLN, KC_QUOT, KC_ENT,
+        KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, KC_COMM, KC_DOT, KC_SLSH, KC_HOME, KC_UP, KC_END,
+        KC_LCTL, KC_LGUI, KC_LALT, KC_SPC, MO(2), MO(1), KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+    [1] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/longnald/corin/keymaps/default/keymap.c:28]:
```suggestion
    [1] = LAYOUT_64_ansi(
```

File: keyboards/longnald/corin/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        KC_GRV, KC_F1, KC_F2, KC_F3, KC_F4, KC_F5, KC_F6, KC_F7, KC_F8, KC_F9, KC_F10, KC_F11, KC_F12, KC_DEL,
+        KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_INS, KC_NO, KC_PSCR, KC_SLCK, KC_PAUS, KC_NO,
+        KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO,
+        KC_LSFT, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_PGUP, KC_NO, KC_PGDN,
+        KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO
+    ),
+    [2] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/longnald/corin/keymaps/default/keymap.c:35]:
```suggestion
    [2] = LAYOUT_64_ansi(
```

File: keyboards/longnald/corin/keymaps/default/keymap.c [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        KC_NO, RGB_TOG, RGB_M_P, RGB_M_B, RGB_M_G, RGB_M_SW, RGB_M_SN, KC_NO, KC_NO, RGB_HUD, RGB_HUI, RGB_VAD, RGB_VAI, KC_NO,
+        KC_NO, KC_NO, LAG_NRM, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO,
+        KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO,
+        KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, LAG_SWP, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO,
+        MO(3), KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO, KC_NO
+    ),
+    [3] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/longnald/corin/keymaps/default/keymap.c:42]:
```suggestion
    [3] = LAYOUT_64_ansi(
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
Renames the layout macro to `LAYOUT_64_ansi`, which opens up the possibility of turning this into a [Community Layout](https://docs.qmk.fm/#/feature_layouts). (There are some other keyboards that support this layout.)

File: keyboards/longnald/corin/info.json (hunk 10 above)
Comment (seatonjiang) [reply keyboards/longnald/corin/info.json:6]: In reference to [info.json Format](https://docs.qmk.fm/#/reference_info_json?id=infojson-format), don't the width and height refer to the size of the pcb?

Comment (noroadsleft) [reply keyboards/longnald/corin/info.json:6]: Yes, but as noted in that doc, it's measured in Key Units. An alpha key is `1`, ANSI Enter is `2.25`, etc. A standard 60% keyboard (like a Poker or DZ60) is `"width": 15, "height": 5`.

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**